### PR TITLE
[Examples] Fixes inconsistency around eval vs val and predict vs test

### LINKED ISSUE #10165: [example scripts] inconsistency around eval vs val
* `val` == validation set (split)
* `eval` == evaluation (mode)

those two are orthogonal to each other - one is a split, another is a model's run mode.

the trainer args and the scripts are inconsistent around when it's `val` and when it's `eval` in variable names and metrics. 

examples:
* `eval_dataset` but `--validation_file`
* `eval_*` metrics key for validation dataset - why the prediction is then `test_*` metric keys?
* `data_args.max_val_samples` vs `eval_dataset` in the same line

the 3 parallels:
- `train` is easy - it's both the process and the split
- `prediction` is almost never used in the scripts it's all `test` - var names and metrics and cl args
- `eval` vs `val` vs `validation` is very inconsistent. when writing tests I'm never sure whether I'm looking up `eval_*` or `val_*` key. And one could run evaluation on the test dataset.

Perhaps asking a question would help and then a consistent answer is obvious:

Are metrics reporting stats on a split or a mode? 
A. split - rename all metrics keys to be `train|val|test`
B. mode - rename all metrics keys to be `train|eval|predict`

Thank you.

@sgugger, @patil-suraj, @patrickvonplaten

### PR DESCRIPTION
**Pull request by bhadreshpsavani**
# What does this PR do?

Fixes #10165 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@stas00 @sgugger

### CODE REVIEW

**Comment by bhadreshpsavani** [comment]
Hi @sgugger and @stas00,
I have made changes in the following way,
| Earlier | Now |
| ---- | ---- |
| test | predict |
| test_examples | predict_examples |
| test_dataset | predict_dataset |
| max_test_samples | max_predict_samples |
| val | eval |
| val_examples | eval_examples |
| val_dataset | eval_dataset |
| max_val_samples | max_eval_samples |

* I have also made changes in the trainer code for the above variables
* I have modified the template file accordingly
* `examples/pytorch/question-answering/run_qa_no_trainer.py` don't have code complete code for predict stage. Shall we need to add?

File: src/transformers/trainer.py
Comment (sgugger) [inline src/transformers/trainer.py:679]: Changing the Trainer like this is a breaking change that we cannot accept (it will break existing code for users relying on the names of the methods or the names of the arguments). This should not be touched by this PR.

File: src/transformers/trainer_pt_utils.py
Comment (sgugger) [inline src/transformers/trainer_pt_utils.py:825]: We should leave `test` as an acceptable value

**Review by sgugger — COMMENTED** [review]
Thanks for your PR, I think there are several more instances where we will need to change `--max_val_samples` and `--max_test_samples`, either in the documentation or the README of examples/pytorch/.

The main blocking point is the changes made in the Trainer: for backward-compatibility reasons we can't change method names or arguments names, so I think you should remove all the changes your made in `trainer.py` (and probably in the trainer-qa to go with it) to limit the changes to examples only.

For the `no_trainer` examples, prediction is fully there, so I'm not sure what you meant?

File: src/transformers/trainer.py
Comment (bhadreshpsavani) [reply src/transformers/trainer.py:679]: Sure @sgugger I will revert these changes

**Comment by bhadreshpsavani** [comment]
Hi @sgugger,
I was trying to test [run_qa_no_trainer.py](https://github.com/huggingface/transformers/blob/master/examples/pytorch/question-answering/run_qa_no_trainer.py) and I came to know that when I passed `--test_file` it was giving me an error since we don't have such [argument](https://github.com/huggingface/transformers/blob/master/examples/pytorch/question-answering/run_qa_no_trainer.py#L87) in the file.

When i give `--do_predict` it gives an error at this [line](https://github.com/huggingface/transformers/blob/master/examples/pytorch/question-answering/run_qa_no_trainer.py#L507) since I was using squad data and it doesn't have `test` set

But as you said it will work perfectly for the dataset with `testset`.

**Comment by sgugger** [comment]
Ah! now I get it. We can add support for a `--test_file` in another PR, yes!

**Comment by bhadreshpsavani** [comment]
I am exploring other files, I will add more changes in docs and other files

**Comment by stas00** [comment]
You're doing a very painful task, @bhadreshpsavani, as @sgugger commented we unfortunately can't normalize these 2 names fully both because of the backcompat and also where in some cases something is called validation/test split :( Thank you for doing this important work and bearing with all these setbacks.

**Comment by bhadreshpsavani** [comment]
Hi @stas00, 
It's totally fine. 
I was expecting more suggestions because I did a lot of changes for all files in a single go. I am enjoying this coding work that's important for me!

**Comment by bhadreshpsavani** [comment]
Hi @stas00,
Can we use this command?
`git push --force-with-lease origin myfeature`
I read that this is safe than `--force` and [many organizations even using this.](https://stackoverflow.com/questions/41283955/github-keeps-saying-this-branch-is-x-commits-ahead-y-commits-behind)

**Comment by stas00** [comment]
In general all PRs are isolated until they are merged, so if you make a mistake on your own PR, in the worst case you will make a mess of your own changes, but it won't impact the master. So feel free to experiment.

I have never used this particular flag before, so if it's safer go for it.

**Comment by stas00** [comment]
`run_tests_torch` CI has been flakey as of recent, you can always force a CI restart if you see one or more CI jobs are failing unrelated to your commit with an empty commit:

```
git commit --allow-empty -m "Trigger CI"
git push
```

**Comment by bhadreshpsavani** [comment]
This is a cool command. I will definitely need this. I will note this down!
Thanks

**Comment by bhadreshpsavani** [comment]
Hi @sgugger,
I have reverted trainer changes and updated the example pytorch readme.
Please let me know if we need to make more changes

**Review by sgugger — APPROVED** [review]
There are two more `max_val_samples` to replace in `tests/deepspeed/test_deepspeed.py`. Apart from that, this is good to merge on my side, thanks for your PR!

**Comment by stas00** [comment]
@sgugger, should these not be covered too?

```
$ grep -Ir max_val_samples
examples/tensorflow/text-classification/run_text_classification.py:    max_val_samples: Optional[int] = field(
examples/tensorflow/text-classification/run_text_classification.py:        if data_args.max_val_samples is not None:
examples/tensorflow/text-classification/run_text_classification.py:            eval_dataset = eval_dataset.select(range(data_args.max_val_samples))
examples/research_projects/wav2vec2/run_common_voice.py:    max_val_samples: Optional[int] = field(
examples/research_projects/wav2vec2/run_common_voice.py:    if data_args.max_val_samples is not None:
examples/research_projects/wav2vec2/run_common_voice.py:        eval_dataset = eval_dataset.select(range(data_args.max_val_samples))
examples/research_projects/wav2vec2/run_common_voice.py:        max_val_samples = data_args.max_val_samples if data_args.max_val_samples is not None else len(eval_dataset)
examples/research_projects/wav2vec2/run_common_voice.py:        metrics["eval_samples"] = min(max_val_samples, len(eval_dataset))
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:    max_val_samples: Optional[int] = field(
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:        if data_args.max_val_samples is not None:
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:            eval_dataset = eval_dataset.select(range(data_args.max_val_samples))
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:            max_val_samples = data_args.max_val_samples if data_args.max_val_samples is not None else len(eval_dataset)
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:            metrics["eval_samples"] = min(max_val_samples, len(eval_dataset))
```

```
$ grep -Ir max_test_samples
examples/tensorflow/text-classification/run_text_classification.py:    max_test_samples: Optional[int] = field(
examples/tensorflow/text-classification/run_text_classification.py:        if data_args.max_test_samples is not None:
examples/tensorflow/text-classification/run_text_classification.py:            test_dataset = test_dataset.select(range(data_args.max_test_samples))
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:    max_test_samples: Optional[int] = field(
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:        if data_args.max_test_samples is not None:
tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py:            test_dataset = test_dataset.select(range(data_args.max_test_samples))
process.txt:--max_train_samples 10 --max_val_samples 10 --max_test_samples 10
```

**Comment by sgugger** [comment]
This PR should be rebased on master and deal with `examples/tensorflow/text-classification/run_text_classification.py` that was added recently yes.

`examples/research_projects/wav2vec2/run_common_voice.py` is a research-project so is not actively maintained and pinned to a version of transformers, I would leave it out of this PR.

For `tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py`, I wouldn't touch it since it's a test script (which will ultimately be replaced by a TF example) but I'll let @philschmid decide on this one.

**Comment by philschmid** [comment]
> For `tests/sagemaker/scripts/pytorch/run_glue_model_parallelism.py`, I wouldn't touch it since it's a test script (which will ultimately be replaced by a TF example) but I'll let @philschmid decide on this one.

Is a custom version for `SageMakerTrainer`, this will be removed after the deprecation of `SageMakerTrainer`, there is no need for adjustments.

**Comment by sgugger** [comment]
Thanks for adjusting! I think this is good to merge, @stas00 if you agree I'll let you click on the button :-)

**Review by stas00 — APPROVED** [review]
Thank you for this massive fixing, @bhadreshpsavani!

**Pull request merged**
